src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/model/Submission.java

```diff
@@ -133,14 +123,40 @@ public Builder setRetry() {
       return this;
     }
 
+
+    public String currentStepName() {
+      if (steps.size() == 0) {
+        return null;
+      }
+      StepResult currentStep = steps.get(steps.size() - 1);
+      if (isComplete(currentStep.getStatus())) {
+        return null;
+      }
+      return currentStep.getName();
+    }
+
     public Builder setStarted() {
-      this.waitTimeMs = System.currentTimeMillis() - dateScheduled.getTime();
-      this.output = null;
+      startStep(WAIT_STEP_NAME);
+      return completeStep(WAIT_STEP_NAME, STATUS_COMPLETE,
+          System.currentTimeMillis() - dateScheduled.getTime(), null);
+    }
+
+    public Builder startStep(String name) {
+      if (this.currentStepName() != null) {
+        throw new IllegalArgumentException();
```

Comment:
Hmm. This is here because it is an error for the caller to interleave steps. So having a checked exception for this seems wrong, because correctly written code can never trigger the exception (and wrong code will always trigger it).